### ISSUE TITLE
Union does not create polygon with holes

### ISSUE THREAD

**Issue by bluenote10** [21re/rust-geo-booleanop#7]
I'm not sure if this a misuse on my side, a problem in the Rust port, or already an issue for the JS implementation. When taking the union of e.g. the following polygons

![debug_polys_input](https://user-images.githubusercontent.com/3620703/71763248-06cbf700-2eda-11ea-8476-e09b7e1c4ad2.png)

I would expect to get 1 polygon with 1 hole. However the output is 2 polygons instead, both having an exterior with empty interiors:

![debug_polys_output](https://user-images.githubusercontent.com/3620703/71763261-2a8f3d00-2eda-11ea-92ae-6329aee2dc13.png)

Code to reproduce:

```rust
extern crate geo;
extern crate geo_booleanop;
extern crate serde;
extern crate serde_json;

use geo_booleanop::boolean::BooleanOp;
use geo::{polygon};
use std::fs::File;


fn main() {

    let polygons = vec![
        polygon![
            (x:   0., y:   0.),
            (x: 100., y:   0.),
            (x: 100., y:  10.),
            (x:   0., y:  10.),
        ],
        polygon![
            (x:  90., y:   0.),
            (x: 100., y:   0.),
            (x: 100., y: 100.),
            (x:  90., y: 100.),
        ],
        polygon![
            (x:   0., y:  90.),
            (x: 100., y:  90.),
            (x: 100., y: 100.),
            (x:   0., y: 100.),
        ],
        polygon![
            (x:   0., y:   0.),
            (x:  10., y:   0.),
            (x:  10., y: 100.),
            (x:   0., y: 100.),
        ],
    ];

    let mut union = geo::MultiPolygon::<f32>(vec![]);
    for poly in &polygons {
        union = union.union(poly);
    }

    let f = File::create("debug_polys_input.json").unwrap();
    serde_json::to_writer_pretty(f, &polygons).unwrap();

    let f = File::create("debug_polys_output.json").unwrap();
    serde_json::to_writer_pretty(f, &union).unwrap();
}
```

Is there a problem how the library interprets holes?

I'm using the latest versions of geo/geo-types/geo-booleanups.

<details>

<summary>JSON output data</summary>

```json
[
  {
    "exterior": [
      {
        "x": 0.0,
        "y": 0.0
      },
      {
        "x": 10.0,
        "y": 0.0
      },
      {
        "x": 90.0,
        "y": 0.0
      },
      {
        "x": 100.0,
        "y": 0.0
      },
      {
        "x": 100.0,
        "y": 10.0
      },
      {
        "x": 100.0,
        "y": 90.0
      },
      {
        "x": 100.0,
        "y": 100.0
      },
      {
        "x": 90.0,
        "y": 100.0
      },
      {
        "x": 10.0,
        "y": 100.0
      },
      {
        "x": 0.0,
        "y": 100.0
      },
      {
        "x": 0.0,
        "y": 90.0
      },
      {
        "x": 0.0,
        "y": 10.0
      },
      {
        "x": 0.0,
        "y": 0.0
      }
    ],
    "interiors": []
  },
  {
    "exterior": [
      {
        "x": 10.0,
        "y": 10.0
      },
      {
        "x": 90.0,
        "y": 10.0
      },
      {
        "x": 90.0,
        "y": 90.0
      },
      {
        "x": 10.0,
        "y": 90.0
      },
      {
        "x": 10.0,
        "y": 10.0
      }
    ],
    "interiors": []
  }
]
```

</details>

<details>
<summary>Python plotting code</summary>

```python
#!/usr/bin/env python

from __future__ import print_function

import matplotlib.pyplot as plt
import json
import sys


if len(sys.argv) != 2:
    print("ERROR: Wrong number of arguments.")
    sys.exit(1)

else:
    filename = sys.argv[1]

    data = json.load(open(filename))

    fig, ax = plt.subplots(1, 1, figsize=(10, 8))

    for i, poly in enumerate(data):
        exterior = poly["exterior"]
        interiors = poly["interiors"]

        xs = [p["x"] for p in exterior]
        ys = [p["y"] for p in exterior]

        ax.plot(xs, ys, "o-", label="Polygon {}".format(i + 1))

    plt.legend(loc="best")
    plt.tight_layout()

    filename_out = filename.replace(".json", ".png")
    plt.savefig(filename_out)

    plt.show()
```

</details>

**Comment by anlumo** [comment]
This is a long-standing issue with the JavaScript version as well. The underlying problem is that the algorithm doesn't handle inner linestrings correctly, so it doesn't map that well to the data structure.

A realistic fix would probably to clean up the result after the operation via some `contains` checks.

**Comment by bluenote10** [comment]
I've actually started looking into many of the open problems of the JS implementation, and hopefully I can provide some solutions soon.

**Comment by anlumo** [comment]
I've created a workaround where I reorder a MultiPolygon to use the proper structuring. It assumes that the LineStrings don't intersect with each other. I hope that I didn't forget any special cases, since there's a lot to consider.

```rust
use geo::{MultiPolygon, Polygon, LineString, Point, CoordinateType, algorithm::contains::Contains};
use std::{
    iter::once,
    cell::RefCell,
    collections::{HashSet, HashMap},
};
use num_traits::Float;

#[derive(Clone)]
struct Contained {
    outers: HashSet<usize>,
    inners: RefCell<HashSet<usize>>,
}

pub(crate) fn fix_inners<N>(mut mpoly: MultiPolygon<N>) -> MultiPolygon<N> where N: CoordinateType + Float {
    let mut strings: Vec<Option<LineString<N>>> = mpoly.0.drain(..).flat_map(|polygon| {
        let (exterior, interiors) = polygon.into_inner();
        interiors.into_iter().chain(once(exterior))
    }).map(|string| Some(string)).collect();

    // generate contained vec + outers
    let mut contained: HashMap<usize, Contained> = strings.iter().enumerate().filter_map(|(idx, line_string)| {
        line_string.as_ref().unwrap().0.first().map(|pt| {
            let pt = Point::new(pt.x, pt.y);
            (idx, Contained {
                // `contains` for LineString only checks whether the point is *on the line*, so we have to convert to a Polygon first.
                outers: strings.iter().enumerate().filter(|(outer_idx, outer)| {
                    *outer_idx != idx && Polygon::new(outer.as_ref().unwrap().clone(), Vec::new()).contains(&pt)
                }).map(|(idx, _)| idx).collect(),
                inners: RefCell::new(HashSet::new()),
            })
        })
    }).collect();

    // generate backlinks (inners)
    for (idx, inner) in &contained {
        for outer_idx in &inner.outers {
            if let Some(outer) = contained.get(&outer_idx) {
                outer.inners.borrow_mut().insert(*idx);
            }
        }
    }

    let mut polygons: Vec<(usize, Contained)> = Vec::new();

    // slice up the line strings like onions, removing one slice from contained at a time
    while contained.len() > 0 {
        let new_polygons: Vec<_> = contained.iter().filter_map(|(idx, outer)| {
            if outer.outers.is_empty() {
                Some((*idx, outer.clone()))
            } else {
                None
            }
        }).collect();

        // remove these polygons and their inner line strings from our list of line strings we still have to handle
        for (idx, polygon) in &new_polygons {
            for (_, contained_inner) in contained.iter_mut() {
                contained_inner.outers.remove(idx);
            }
            let not_direct_child: Vec<_> = polygon.inners.borrow().iter().filter(|inner_idx| contained.get(&inner_idx).map(|inner| !inner.outers.is_empty()).unwrap_or(true)).cloned().collect();
            for inner in not_direct_child {
                polygon.inners.borrow_mut().remove(&inner);
            }

            // remove the outer line strings (that became a polygon) and their inner line strings
            contained.retain(|_, inner| !inner.outers.is_empty());
        }
        for (_, contained_inner) in contained.iter_mut() {
            for (_, polygon) in &new_polygons {
                for inner_idx in polygon.inners.borrow().iter() {
                    contained_inner.outers.remove(inner_idx);
                }
            }
        }

        polygons.extend(new_polygons);
    }

    MultiPolygon(polygons.into_iter().map(|(idx, contained)| Polygon::new(strings[idx].take().unwrap(), contained.inners.into_inner().into_iter().filter_map(|idx| strings[idx].take()).collect())).collect())
}
```

**Comment by bluenote10** [comment]
I think I've figured out how it is supposed to work. 

@anlumo Would be interesting to test your use case after #10 is merged/released to see if it also works on your data.

**Comment by bluenote10** [comment]
Solved by #10
